### PR TITLE
WIP: Riscv 0.13 Working on GD103VF103, debug working on ESP32-C3

### PR DESCRIPTION
**Pull request by UweBonnes**
Disclaimers:
- May ruin your debugger or target
- Forced push may happen

Prerequisite:
- For hosted/firmware use, compile and flash firmware  from this branch.
- Firmware should still be compatible with older hosted software

What should work:
- Core/Hart Detection and debugging
- Memory setup on GD32VF103 and flashing

What does not work (yet?)
- ESP32-C3 memory setup, help needed
- ESP32-C3 flashing: Probably needs major effort

Hint:
- Use hosted for verbose output

Speed :
- GD32FV103/firmware ~ 40 kByte/s read ~10 kByte/s write
- GD32FV103/remote ~ 9kByte/s read ~3 kByte/s write
- GD32FV103/builtin CMSIS-DAP on GD32VF103C-START: ~ 2kByte/s.
- ESP32-C3, FT2232H adapter: ~ 27 kByte/s

### CODE REVIEW

**Comment by ruabmbua** [comment]
Is there a diff to my original work? Would make it simpler for me to remember.

**Comment by UweBonnes** [comment]
To much work done in between. I tested with GD32VF and ESP32-C3 and both did not work with progbuf  memory access.

File: src/arch/riscv/rvdbg013.c [hunk 1]
```diff
@@ -0,0 +1,1772 @@
...
+	}
+	uint32_t data;
+	rvdbg_dmi_read(dmi, DMI_REG_ABSTRACT_CS, &data);
+	if (ABSTRACTCS_GET_CMDERR(data)) {
+		DEBUG_WARN("mem_read_abstract failure\n");
+		dmi->error = true;
+	}
+}
+
+static void rvdbg_mem_read_systembus(target *t,  void* dest, target_addr address, size_t len)
```
Comment (ruabmbua) [inline src/arch/riscv/rvdbg013.c:1009]: maybe the different mem read & write implementations should multiplex at the byte & word r/w functions like rvdbg_mem_read_systembus, and generalize the repeated align to word + copy the rest in words logic?

File: src/arch/riscv/rvdbg013.c [hunk 2]
```diff
@@ -0,0 +1,1772 @@
...
+		DEBUG_WARN("Poll write dmcontrol failed\n");
+	if (!DMSTATUS_GET_ALLHALTED(dmstatus))
+		return TARGET_HALT_RUNNING;
+	/* dcsr may not be readable when running*/
+	uint32_t dcsr;
+	res = rvdbg_read_single_reg(dmi, HART_REG_CSR_DCSR, &dcsr, AUTOEXEC_STATE_NONE);
+	uint8_t cause = (dcsr >> 6) & 7;
+	if ((t->t_designer == 0x612) && (cause == 3 ) && (dcsr & CSR_DCSR_STEP)) {
+		/* FIXME: ESP32-C3 never reports TARGET_HALT_STEPPING */
+		DEBUG_INFO("Workaround for single stepping ESP32-C3\n");
```
Comment (ruabmbua) [inline src/arch/riscv/rvdbg013.c:1360]: is it possible to identify the esp32-c3? maybe implement this explicitly as a quirk, makes it more visible, it is a target specific workaround. Other hw should probably still be handled normally.

File: src/arch/riscv/rvdbg013.c [hunk 3]
```diff
@@ -0,0 +1,1772 @@
...
+			DEBUG_INFO("leave run/test immediately\n");
+			break;
+		default:
+			DEBUG_INFO("stay %d cycles in run/test\n", dmi->idle - 1);
+			break;
+	}
+
+	dmi->rvdbg_dmi_reset(dmi, false);
+	/* Reset DM to initial values.
+	 * 0,13 to 1.0 incompatible change: Poll dmactive after lowering it. #566
```
Comment (ruabmbua) [inline src/arch/riscv/rvdbg013.c:1471]: we already know the spec version here, maybe do the poll optionally depending on the reported version?

File: src/target/jtag_devs.c [hunk 4]
```diff
@@ -57,5 +61,6 @@ jtag_dev_descr_t dev_descr[] = {
 	{.idcode = 0x20270013, .idmask = 0xFFFFFFFF, .descr = "Intel: i80386ex."},
 	{.idcode = 0x07B7617F, .idmask = 0xFFFFFFFF, .descr = "Broadcom: BCM2835."},
 	{.idcode = 0x4BA00477, .idmask = 0xFFFFFFFF, .descr = "Broadcom: BCM2836."},
+	{.idcode = 0x790007A3 , .idmask = 0xFFFFFFFF, .descr = "GigaDevice: GD32VF103CBT6 (BSD)."},
```
Comment (ruabmbua) [inline src/target/jtag_devs.c:64]: Is there actually a second GD32VF idcode?, or was this perhaps a copy & paste mistake?

File: src/arch/riscv/rvdbg013.c (hunk 2 above)
Comment (UweBonnes) [reply src/arch/riscv/rvdbg013.c:1360]: It is already found:
...
Machine 612, 80000001, 1, 0
Found 8 triggers
***  1      ESP32-C3 Generic RVDBG 0.13

File: src/arch/riscv/rvdbg013.c [hunk 5]
```diff
@@ -0,0 +1,1772 @@
...
+		abstractauto = 0;
+		ABSTRACTAUTO_SET_DATA(abstractauto, 0);
+		if (rvdbg_dmi_write(dmi, DMI_REG_ABSTRACT_AUTOEXEC, abstractauto) < 0)
+			return -1;
+	}
+
+	return err;
+}
+
+static void rvdbg_regs_read(target *t, void *regs_data)
```
[the remark this answers is not in the document]
Comment (UweBonnes) [reply src/arch/riscv/rvdbg013.c:692]: Both reg_read|write and regs_read|write can be requests from gdb.

**Comment by UweBonnes** [comment]
I successfully flash some GD32VF platformio project via GDB and was able to do run control and breakpoints. With BMP firmware, write speed was about 10 kByte / s and read speed 40 kByte/s

**Comment by fabalthazar** [comment]
Here is a battery of some tests performed on 557f938 (Wed Oct 27 11:48:39 2021 +0200)
Target: GD32VF103 (Sipeed Longan nano).
BMP platform: Blue pill (swlink).
main() function used:
```
int main(void)
{
    volatile uint32_t a = 0U;
    longan_led_init();

    while(1){
        /* turn on builtin led */
        longan_led_on();
        a++;
        delay_1ms(1000);
        /* turn off uiltin led */
        longan_led_off();
        if (a == 10U)
            a = 0U;
        delay_1ms(1000);
    }
}
```

**Issue 1**: The target is incorrectly recognized as Risc-V 0.11 from hosted `blackmagic -t` if the probe has outdated firmware (e.g. <= efa8891 - Tue Nov 16 14:23:20 2021 +0100).
Output:
```shell
$ blackmagic -t
...
Speed set to  3.2727 MHz for JTAG
Update old firmware for faster JTAG
 1: IDCODE: 0x1000563d, IR len 5 RVDBG013
 2: IDCODE: 0x790007a3, IR len 5 GIGADEVICE BSD (Unhandled) 
RISC-V DTM id 0x1000563d detected: `RVDBG013`
Scanning RISC-V target ...
  dtmcs = 0x00000000
Error: RISC-V debug 0.11 not supported
Can not attach to target 1
```

Once the firmware is updated, the following tests passed:
- Probing from both hosted and GDB;
- Flashing from both hosted and GDB;
- Erase by blocks from hosted (blackmagic -E);
- mass_erase from monitor command;
- Pause and resume;
- Breakpoints.

However, a case about watchpoints need further test and investigation.
**Issue 2**:
- Place a watchpoint on the variable `a`. For instance `watch a` or `awatch a`;
- Continue;
- The target will successfully break on any instruction involving `a` (for instance `a++;`);
- Continue without touching the breakpoint configuration;
- The target camps on the same instruction and thus after many `continue` requests.

**Issue 3**:
- Perform the same test as issue 2 until the target halts on watchpoint;
- Instead of `continue`, perform a stepping with `next` request;
- GDB hangs and needs to be terminated.

Perhaps issues 2 and 3 are linked by the same cause. As if a watchpoint-related flag not cleared prevents the CPU from continuing executing next machine instructions. Just a thought.

To be tested (non exhaustive):
- Debug in low power mode(s);
- Connect under reset;
- ...

**Comment by fabalthazar** [comment]
The watchpoint bug identified as issues 2 and 3 above has been understood.
A watchpoint hit is currently notified to the gdb client as a breakpoint.
The lines of interest are located in the rvdbg_halt_poll() function:
```
switch (cause) {
	case 1: /* Software breakpoint */
	case 2: /* Hardware trigger breakpoint */
		return TARGET_HALT_BREAKPOINT;
```
The `cause` being the cause bitfield of the DCSR Risc-v register informing about the halt reason. This one does not distinguishes a watch/breakpoint by itself and may need a further evaluation of the root break/watchpoint responsible for the halt.

Documentation on the Debug Control and Status Register (dcsr) is located in chapter 4.8.1 of the RISC-V External Debug Support, version 0.13.2.

**Comment by fabalthazar** [comment]
Just an update to track the work done on RISC-V.

I have worked on fixing the watchpoint issue above-mentioned. Currently in my 'ruabmbua' branch.
The missing feature was to figure out what is the watched address when the watchpoint is hit. The RISC-V debug architecture provides the _optional_ bit for that. As GD32VF103 (sadly) does not implement this bit, I have implemented the instruction decoding approach (also done in rv-link). Nevertheless, I plan to implement the 'hit' method for supported devices (maybe ESP32-C3?).

In the meantime, I found another issue about target memory reading and especially 32-bit words. Let's show an example.

**Issue 4**:

I know for sure that the half word 0x47b2 resides at address 0x08000436 (my example program is as is) and here is what I read from GDB:
```
(gdb) x /wx 0x08000436
0x8000436 <main+18>:    0x656a0785
(gdb) x /hx 0x08000436
0x8000436 <main+18>:    0x47b2
```
The 32-bit word read is erroneous.
If I perform separate 8-bit readings:
```
(gdb) x /bx 0x08000436
0x8000436 <main+18>:    0xb2
(gdb) 
0x8000437 <main+19>:    0x47
(gdb) 
0x8000438 <main+20>:    0x85
(gdb) 
0x8000439 <main+21>:    0x07
```
Which confirms the right 0x47b2 but also shows 0x0785 which seems wrongly read as least significant bits on the previous 32-bit reading.
Reading 0x08000438 as an half word confirms 0x0785:
```
(gdb) x /hx 0x08000438
0x8000438 <main+20>:    0x0785
```
Additionally, the basic printf() approach on `target_mem_read32(t, 0x08000436)` gives 0x00000785 which is none of previous reads! 0x0785 is still wrongly there.
A proper fix of this 32-bit memory reading is needed in order to, on top of bringing this essential feature, test the instruction decoding on usual RV32I 32-bit instructions. I tested on RVC compressed 16-bit ones.

I need more time to understand and fix this. Any help is welcome :).

**Comment by UweBonnes** [comment]
It is a pity that the Riscv spec has so many "may", "can", "optional". Together with v0.11, v0.13 and now v1 possibilities are near endless.

**Comment by koendv** [comment]
Hi. A question: how do you attach to the following risc-v processor (huawei/hisilicon hi3861v100)

[hi3861.txt](https://github.com/blackmagic-debug/blackmagic/files/7936321/hi3861.txt)

File: src/platforms/hosted/cmsis_dap.c [hunk 6]
```diff
@@ -402,6 +402,8 @@ int cmsis_dap_jtagtap_init(jtag_proc_t *jtag_proc)
 	jtag_proc->jtagtap_tms_seq     = cmsis_dap_jtagtap_tms_seq;
 	jtag_proc->jtagtap_tdi_tdo_seq = cmsis_dap_jtagtap_tdi_tdo_seq;
 	jtag_proc->jtagtap_tdi_seq     = cmsis_dap_jtagtap_tdi_seq;
+	jtag_proc->dev_shift_ir        = dap_shift_ir;
```
Comment (DX-MON) [inline src/platforms/hosted/cmsis_dap.c:405]: These changes and the ones in dap.h/c feel like they belong in a separate PR with the jtagtap.c/h changes

File: src/platforms/hosted/platform.c [hunk 7]
```diff
@@ -40,7 +40,13 @@
 
 bmp_info_t info;
 
-jtag_proc_t jtag_proc;
```
Comment (DX-MON) [inline src/platforms/hosted/platform.c:43]: These changes too feel like they belong with the jtagtap.c/h and dap.c/h changes seperately

File: src/platforms/stm32/gdb_if.c [hunk 8]
```diff
@@ -24,6 +24,7 @@
  */
 #include "general.h"
 #include "cdcacm.h"
+#include <libopencmsis/core_cm3.h>
```
Comment (DX-MON) [inline src/platforms/stm32/gdb_if.c:27]: These changes feel like a separate STM32 platforms bugfix PR

**Comment by UweBonnes** [comment]
This PR is labeled "WIP" (work in progress) and exposed so that people can test and contribute. It is work of several people, Ruabmua, Fabalthazar, Gareth and me to name some over several years. It is not intended to be merged as it. Some things helped local setup, some bugs and their fixes only get exposed when using this branch, some fixes needs a check if  right by getting coverage also in other use cases and some fixes may have another solution already.  What is general useful will get its own PR and things will get a cleanup before the WIP tag is remove. But only usage by other users is best to enhance usefulness.

**Comment by esden** [comment]
I think the point that @DX-MON is makeing is that it would be good to start moving parts of the changes this PR depends on into separate PRs. If we start doing that, this PR will become a smaller and more focused on RISC-V support. This way we will actively make a move from this being a WIP and being able to merge it eventually. I bet this is the goal of everyone involved with this PR.

**Comment by UweBonnes** [comment]
@koendv: Probably you can not attach in the present state to a hi386. The riscv debug spec  offers many "can" , "may" and alternatives and together with V0.11, V0.13 and now V1 possibilities are endless. You only show a openocd log, but probably that is a special build, as neither "hi386" nor "Hi-ft2232d-ftdi" give a hit in the openocd codebase.  Can you try with the ruabmbua branch? If you point me to Hi-ft2232d-ftdi.cfg and best also the schematics, I can add  or help you to add an  entry for that cable in ftdi_bmp.c so you can also run blackmagic/hosted on your setup unchanges

**Comment by fabalthazar** [comment]
The last commits that slipped into this PR are supposed to fix what I named "Issues 2,3,4" before.
Issue 1 is still pending ("Error: RISC-V debug 0.11 not supported" message when firmware outdated and actual v0.13 enabled target).

That said, I am interested in a test report on a non GD32VF103 device and some watchpoints set.
I'd like to here about a chip that implements the straightforward 'hit' bit in the mcontrol register to signal the right watchpoint hit.
Does someone has an ESP32-C3 or other and can place let's say 2 watchpoints of any type (watch/rwatch/awatch)?
Preferably, use hosted with "TARGET" debug messages enabled : `blackmagic -j -v5`, play with watchpoints and look the blackmagic (very verbose) output for any "hit bit unimplemented" message. If this is absent, it means the hit bit method works well and we found a friendly chip.

**Comment by koendv** [comment]
The hisilicon hi3681 is v0.13, I think. I've added a picture so it's easier to know what we're talking about.
The hi3861 is a small microcontroller; 2mb flash, 384 kbyte ram, up to 160 MHz.
Wifi, bluetooth. Runs [Huawei LiteOS](https://gitee.com/LiteOS/). 

I've added openocd scripts and schematics. The openocd scripts connect using an ft2232. The schematics show how to enable jtag. 

hi3861
[hi3681.zip](https://github.com/blackmagic-debug/blackmagic/files/7963949/hi3681.zip)

**Comment by koendv** [comment]
On Tue, 25 Jan 2022 12:36:03 -0800
UweBonnes ***@***.***> wrote:

> If you point me to Hi-ft2232d-ftdi.cfg and best also the schematics

Schematics and openocd config files are in the file hi3861.zip above.

I've taken a look at why rtt fails on GD32VF103.

On arm cortex-m, the debugger can read memory via jtag while the target is running.
On risc-v, memory access via jtag while the target is running is not guaranteed. (par. 3.7.1.3).

What does this mean for rtt?
In the most simple form, rtt is a circular buffer in target memory, written by the target and read by the debugger.
For rtt on risc-v, not being able to read memory while the target is running, means halting and resuming the target every time rtt in the debugger wants to access the circular buffers on the target. This makes use of rtt on risc-v less appealing.

koen

**Comment by koendv** [comment]
Problem: target hangs if debugger writes a single byte to target.
```
koen@raspberrypi:~/src/riscv/gd32vf103inator $ /opt/riscv-binutils-gdb/bin/riscv-elf-gdb -q
(gdb) tar ext /dev/ttyACM0
Remote debugging using /dev/ttyACM0
(gdb) mon ver
Black Magic Probe(F4Discovery/BlackPillV2) v1.6.1-977-g5d40c88-dirty, Hardware Version 0
Copyright (C) 2015  Black Sphere Technologies Ltd.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

(gdb) mon jtag
Available Targets:
No. Att Driver
 1      GD32VF1 Generic RVDBG 0.13
(gdb) att 1
Attaching to Remote target
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x08000486 in ?? ()
(gdb) file build/code.elf 
A program is being debugged already.
Are you sure you want to change the file? (y or n) y
Reading symbols from build/code.elf...
(gdb) p ch
$1 = 62 '>'
(gdb) p ch='a'
Ignoring packet error, continuing...
Ignoring packet error, continuing...
Ignoring packet error, continuing...
```
Apart from that, rtt output from riscv target to host seems to work.

**Comment by fabalthazar** [comment]
How is that possible that your probe firmware has old version v1.6.1 and it speaks risc-v? I have not found any g5d40c88 commit in the risc-v tree.

For assigning target variables, I tend to use `set var=value` GDB syntax but I guess you've seen `p var=value` working as well.

**Comment by koendv** [comment]
This is what I do to build bmp firmware with riscv and rtt:
```
git clone -b ruabmbua https://github.com/UweBonnes/blackmagic
wget -O rtt.patch https://github.com/blackmagic-debug/blackmagic/commit/c3c4ad99a6407024cae6de59a83c8a30fd9748e3.patch
cd blackmagic
patch -p1 < ../rtt.patch
make PROBE_HOST=f4discovery BLACKPILL=1 ENABLE_RTT=1
```
Should I try another git instead?

**Comment by fabalthazar** [comment]
My mistake, you're well on risc-v head, I forgot to remove the heading 'g' from the git revision: `git log --oneline | grep 5d40c88`.
I see, you may have missed some tags by cloning, so the automatic version generation is out of sink (`git describe --always --dirty --tags` should show 1.6.1 in your case, instead of 1.7.1). Doesn't matter so much since we know 5d40c88 is the actual head.
You seem to be all correct.

**Comment by koendv** [comment]
I've repeated the experiment, this time with plain  'git clone -b ruabmbua https://github.com/UweBonnes/blackmagic', bmp on a f4discovery/blackpill.
```
 $ /opt/riscv-binutils-gdb/bin/riscv-elf-gdb -q
(gdb) tar ext /dev/ttyACM0
Remote debugging using /dev/ttyACM0
(gdb) mon jtag
Available Targets:
No. Att Driver
 1      GD32VF1 Generic RVDBG 0.13
(gdb) att 1
Attaching to Remote target
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x08000486 in ?? ()
(gdb) set {int}0x20000000=1
(gdb) set {char}0x20000000=1
Ignoring packet error, continuing...
Ignoring packet error, continuing...
Ignoring packet error, continuing...
```
If I repeat the last command with debug serial enabled,
you can clearly see bmp hangs:
```
(gdb) set debug serial 1
(gdb) set {char}0x20000000=1
[
w $][X][2][0][0][0][0][0][0][0][,][1][:][\x01][#][7][2][
r +][<Timeout: 2 seconds>][
w -][
r <Timeout: 2 seconds>][
w -][
r <Timeout: 2 seconds>][
w -]Ignoring packet error, continuing...
```

**Comment by koendv** [comment]
If I rebase this to main, it does not compile.

**Comment by koendv** [comment]
OK. What am I doing wrong?
```
git clone -b feature/risc-v https://github.com/blackmagic-debug/blackmagic
git remote add bmp https://github.com/blackmagic-debug/blackmagic
git fetch bmp
git merge-base feature/risc-v bmp/main
1ea9641adaa4b6e7a767ff970ffac328d9b3b5b8
git rebase 1ea9641adaa4b6e7a767ff970ffac328d9b3b5b8
make PROBE_HOST=blackpillv2 ENABLE_RTT=1
target/riscv0_13.c:1419:15: error: 'target' {aka 'struct target_s'} has no member named 't_designer'
 1419 |         if ((t->t_designer == 0x612) && (cause == 3 ) && (dcsr & CSR_DCSR_STEP)) {
      |               ^~
```

**Comment by dragonmux** [comment]
Closing this PR in favour of PR #1380 on account of how much work is required to bring this up to speed and the effort required to sort what's here out and get it complying with the contribution guidelines
